packages/@aws-cdk/aws-apigateway/lib/authorizers/lambda.ts

```diff
@@ -0,0 +1,112 @@
+   */
+  readonly function: lambda.IFunction;
+
+  /**
+   * The name of the header in the request that contains the authorization token as submitted by the client.
+   */
+  readonly headerName: string;
+
+  /**
+   * The TTL on how long APIGateway should cache the results. Max 1 hour.
+   * Disable caching by setting this to 0.
+   *
+   *  @default - 300 seconds
+   */
+  readonly resultsCacheTtl?: Duration;
+
+  /**
+   * An optional regex to be matched against the authorization token. When matched the authorizer lambda is invoked,
+   * otherwise a 401 Unauthorized is returned to the client.
+   *
+   * @default - no regex filter will be applied.
+   */
+  readonly validationRegex?: string;
+
+  /**
+   * An optional IAM role for APIGateway to assume before calling the Lambda-based authorizer.
+   *
+   * @default - A resource policy is added to the Lambda function allowing apigateway.amazonaws.com to invoke the function.
+   */
+  readonly assumeRole?: iam.IRole;
+}
+
+/**
+ * Token based lambda authorizer that recognizes the caller's identity caller's identity as a bearer token,
+ * such as a JSON Web Token (JWT) or an OAuth token.
+ * Based on the token, authorization is performed by a lambda function.
+ *
+ * @resource AWS::ApiGateway::Authorizer
+ */
+export class TokenAuthorizer extends CustomAuthorizer {
+
+  /**
+   * The id of the authorizer.
+   * @attribute
+   */
+  public readonly authorizerId: string;
+
+  /**
+   * The ARN of the authorizer to be used in permission policies, such as IAM and resource-based grants.
+   * @attribute
+   */
+  public readonly authorizerArn: string;
+
+  constructor(scope: Construct, id: string, props: TokenAuthorizerProps) {
+    super(scope, id, {
+      physicalName: props.name || Lazy.stringValue({ produce: () => this.node.uniqueId })
+    });
+
+    if (props.resultsCacheTtl && props.resultsCacheTtl.toSeconds() > 3600) {
+      throw new Error(`Lambda authorizer property 'cacheTtl' must not be greater than 3600 seconds`);
+    }
+
+    const resource = new CfnAuthorizer(this, 'Resource', {
+      name: this.physicalName,
+      restApiId: this.restApiId,
+      type: 'TOKEN',
+      authorizerUri: `arn:aws:apigateway:${Stack.of(this).region}:lambda:path/2015-03-31/functions/${props.function.functionArn}/invocations`,
+      authorizerCredentials: props.assumeRole ? props.assumeRole.roleArn : undefined,
+      authorizerResultTtlInSeconds: props.resultsCacheTtl && props.resultsCacheTtl.toSeconds(),
+      identitySource: `method.request.header.${props.headerName}`,
+      identityValidationExpression: props.validationRegex,
+    });
+
+    this.authorizerId = resource.ref;
+
+    this.authorizerArn = Stack.of(this).formatArn({
+      service: 'execute-api',
+      resource: this.restApiId,
+      resourceName: `authorizers/${this.authorizerId}`
+    });
+
+    if (!props.assumeRole) {
+      new lambda.CfnPermission(this, `${this.node.uniqueId}:Permissions`, {
```

Comment:
We really try to avoid using L1s across modules (consider L1s "private"). Use `lambda.addPermission` instead and if it's not good enough, let's improve it.

Also, wouldn't it make sense that if user specifies a role this role will be granted permission to invoke the lambda?

Reply:
> We really try to avoid using L1s across modules (consider L1s "private").

Duh. Don't know why I didn't do that.